core/src/header_chain_prover.rs

```diff
@@ -0,0 +1,769 @@
+where
+    R: RpcApiWrapper,
+{
+    rpc: ExtendedRpc<R>,
+    db: Database,
+}
+
+impl<R> HeaderChainProver<R>
+where
+    R: RpcApiWrapper,
+{
+    pub async fn new(config: &BridgeConfig, rpc: ExtendedRpc<R>) -> Result<Self, BridgeError> {
+        let db = Database::new(config).await?;
+
+        if let Some(proof_file) = &config.header_chain_proof {
+            let file = File::open(proof_file).map_err(|e| {
+                BridgeError::ProveError(format!(
+                    "Can't read assumption file {:?} with error {}",
+                    proof_file, e
+                ))
+            })?;
+            let mut reader = BufReader::new(file);
+            let mut assumption = Vec::new();
+            reader.read_to_end(&mut assumption)?;
+
+            let proof: Receipt = borsh::from_slice(&assumption).map_err(|e| {
+                BridgeError::ProveError(format!("Proof assumption is malformed: {}", e))
+            })?;
+            let prev_output: BlockHeaderCircuitOutput = borsh::from_slice(&proof.journal.bytes)
+                .map_err(|e| {
+                    BridgeError::ProveError(format!("Can't convert journal to bytes: {}", e))
+                })?;
+
+            // Create block entry, if not exists.
+            let block_hash = rpc
+                .client
+                .get_block_hash(prev_output.chain_state.block_height.into())?;
+            let block_header = rpc.client.get_block_header(&block_hash)?;
+            // Ignore error if block entry is in database already.
+            let _ = db
+                .save_new_block(
+                    None,
+                    block_hash,
+                    block_header,
+                    prev_output.chain_state.block_height.into(),
+                )
+                .await;
+
+            // Save proof assumption.
+            db.save_block_proof(None, block_hash, proof).await?;
+        };
+
+        Ok(HeaderChainProver { rpc, db })
+    }
+
+    /// Get the proof of a block.
+    ///
+    /// # Parameters
+    ///
+    /// - `hash`: Target block hash
+    pub async fn get_header_chain_proof(&self, hash: BlockHash) -> Result<Receipt, BridgeError> {
+        match self.db.get_block_proof_by_hash(None, hash).await? {
+            Some(r) => Ok(r),
+            None => Err(BridgeError::ProveError(format!(
+                "No proof is present for block with block hash {}",
+                hash
+            ))),
+        }
+    }
+
+    /// Starts a background task that syncs current database to active
+    /// blockchain and does proving.
+    #[tracing::instrument]
+    pub fn start_header_chain_prover(&self) {
+        let (tx, rx) = mpsc::channel::<()>(5);
+
+        // Block checks.
+        let block_checks = HeaderChainProver {
+            rpc: self.rpc.clone(),
+            db: self.db.clone(),
+        };
+        let block_gazer = HeaderChainProver::start_blockgazer(block_checks, tx);
+
+        // Prover.
+        let prover = HeaderChainProver {
```

Comment:
is this not clonable

Reply:
This is needed because of `ExtendedRpc`'s `RpcApiWrapper`:

```
`Clone` is not implemented because the following trait bounds could not be satisfied: `_: Sized`, `_: Clone` and `_: RpcApiWrapper`
```

Once mock_rpc is ditched, this can be fixed: Added a TODO.